### PR TITLE
feat(reports): Programmatic Report Delivery via Webhook

### LINKED ISSUE #11517: [FEATURE] - [Reports] Add Programmatic Report Delivery via Webhook
### Feature Description

**Current State and Problem:** 
Currently, our asynchronous report generation delivers the final S3 download link via email. While this prevents API timeouts, it restricts merchants from building true machine-to-machine automation. They are forced to rely on "human-in-the-loop" processes or email-parsing scripts to retrieve their data.

**The Solution:**
We need to upgrade the report generation flow to support true programmatic delivery via webhooks.
Merchants will be able to pass a return_url in their initial API request. Once the background Lambda worker finishes generating the report, it will fire an HTTP POST webhook to this URL.

**Key Business & UX Benefits:**
* **True Automation:** Merchants can trigger a report via API and have their backend automatically ingest the resulting file minutes later without any human intervention.
* **Structured Payload:** The webhook will deliver a standardized JSON payload with merchant_id and org_id hoisted to the root level, making enterprise routing effortless for our clients.

### Possible Implementation

**1. API Payload Modification (Rust):**
Update the `ReportRequest` struct to accept the new webhook URL.
* Add `pub return_url: Option<Url>`.
* Use the `#[serde(default)]` macro. This guarantees backward compatibility for clients still relying on the legacy email flow (if they omit the field, it safely defaults to `None`).

**2. Application-Layer URL Validation (Rust)**
Although our infrastructure layer validates all URLs against a whitelist, we must still implement strict return_url validation at the application layer.
Before accepting the request, the Rust API must:
* **Enforce HTTPS:** Reject `http://` schemes.
* **Resolve DNS (`to_socket_addrs`):** Convert the host to its underlying IP addresses before validation to prevent from masking internal IPs behind public domain names.
* Block Private IPs: Iterate through the resolved IPs and fail fast (`400 Bad Request`) if any resolve to Localhost (`127.0.x`, `0.0.0.0`), internal VPC networks (`10.x, 172.16.x, 192.168.x`), or AWS Cloud Metadata (`169.254.x.x`).

**Benefits of Application Layer URL Validation:**
* **Cost Savings & UX (Failing Fast):** Validating the URL upfront means we immediately reject invalid or restricted IPs with a `400 Bad Request`. This prevents us from unnecessarily spinning up a Lambda worker, executing expensive database queries, and generating a report that the infrastructure would just end up blocking anyway. It saves compute costs and gives the merchant immediate feedback.

### Have you spent some time checking if this feature request has been raised before?

- [x] I checked and didn't find a similar issue

### Have you read the Contributing Guidelines?

- [x] I have read the [Contributing Guidelines](https://github.com/juspay/hyperswitch/blob/main/docs/CONTRIBUTING.md)

### Are you willing to submit a PR?

Yes, I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by rohanm13**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [x] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

This PR upgrades our asynchronous report generation to support true programmatic delivery. Merchants can now provide a `returnUrl` in their API request, and our background worker will fire a webhook with the S3 download link when the report is ready, eliminating the need for manual email parsing.

### Additional Changes

- [x] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

**The Product Problem: Breaking the Automation Bottleneck**
Historically, our asynchronous report generation delivered the final S3 download link via email. While this successfully prevented API timeouts for massive datasets, it created a massive bottleneck for our merchants. To automate their data ingestion, they were forced to write fragile email-parsing scripts or rely on a human to manually click links.
By introducing the `returnUrl` parameter and webhook delivery, we are unlocking true machine-to-machine automation. Merchants can now programmatically trigger a report and have their backend automatically ingest the data minutes later via a structured JSON payload, providing a modern, enterprise-grade developer experience.

**The Engineering Context: Why Application-Layer Validation?**
To support this webhook feature safely, we had to introduce strict URL validation in the Rust API layer before the request is handed off to the Lambda worker.
By immediately rejecting private IPs or unresolvable domains with a `400 Bad Request`, we prevent AWS from spinning up a Lambda and executing expensive database queries for a webhook that can never be securely delivered.

## How did you test it?

I manually tested the API endpoint against various edge cases to ensure the application-layer SSRF protection works exactly as intended, while preserving backward compatibility.
**1. Backward Compatibility (Legacy Flow)**
* **Input:** Submitted a `POST /reports_request` omitting the `return_url` field entirely.
* **Result:** Payload parsed successfully (`return_url` defaulted to `None`). The background worker successfully generated the report and did the email delivery flow.

**Request:**
```bash
{
    "timeRange": {
        "startTime": "2026-02-06T18:30:00Z",
        "endTime": "2026-03-09T05:36:07Z"
    },
    "emails": ["rohan.mali@juspay.in"]
}
```
**Response**
```bash
null
```
**2. Webhook Delivery Functionality**
* **Input:** Provided a valid public webhook URL (e.g., `https://webhook.site/...`).
* **Result:** API returned `200 OK`.

**Request**
```bash
{
    "timeRange": {
        "startTime": "2026-02-06T18:30:00Z",
        "endTime": "2026-03-09T05:36:07Z"
    },
    "emails": ["rohan.mali@juspay.in"],
    "returnUrl": "https://google.com"
}
```
**Response**
```bash
null
```

### Worker Dispatcher & Payload Formatting (Isolated Mock Test)
* **Context:** Because the development environment's egress firewall restricts outbound traffic to a strict whitelist, I could not test a full end-to-end webhook delivery to a random URL.
* **Method:** I tested the Lambda worker's dispatch logic in isolation. I temporarily hardcoded the expected S3/Report data and passed a public `https://webhook.site/bee7a938-76ed-42d4-a0a9-0d4e588af51d` URL as the destination.
* **Result:** The Lambda successfully constructed the structured JSON payload and executed the HTTP POST request. The data arrived at the webhook site exactly as expected, proving the `urllib` dispatcher, timeout handling, and JSON serialization work perfectly. (See attached screenshot of the webhook payload).
```bash
{
    "event_type": "report_generation.completed",
    "status": "success",
    "org_id": "org_abcdefghijklmnop",
    "merchant_id": "merchant_1770815665",
    "data": {
       "report_type": "authentication_report",
       "start_date_utc": "2026-03-08",
       "end_date_utc": "2026-03-13",
       "download_url": "https://mock-bucket.s3.amazonaws.com/test_report.csv",
       "expires_in_hours": 48
    }
}
```

### End-to-End Programmatic Execution (Instructions for QA)
To verify this feature in Integ/Sandbox environment, the entire pipeline must be triggered programmatically to simulate a real merchant.
* **Method:** Trigger the `POST https://app.hyperswitch.io/api/analytics/v1/merchant/report/payments` endpoint via cURL using a valid API Key in the `Authorization` header.
* **Crucial Infrastructure Note:** The `returnUrl` provided in this E2E test **must already be on our infrastructure's egress whitelist**. If a random URL is used, the Rust API will accept it, the Lambda will process it, but the egress firewall will intentionally drop the final payload.
* **Expected Result:** The system authenticates the request, validates the URL, fires the Lambda, and the infrastructure allows the final JSON payload to reach the whitelisted destination.
```bash
curl --location 'https://app.hyperswitch.io/api/analytics/v1/merchant/report/payments' \
--header 'X-Merchant-Id: <MERCHANT_ID>' \
--header 'X-Profile-Id: <MERCHANT_PROFILE_ID>' \
--header 'api-key: <MERCHANT_API_KEY>' \
--header 'Content-Type: application/json' \
--data-raw '{
    "timeRange": {
        "startTime": "2026-02-06T18:30:00Z",
        "endTime": "2026-03-09T05:36:07Z"
    },
    "emails": ["abcd@example1.com", "pqrs@example2.com"],
    "returnUrl": "https://whitelisted-test-domain.com"
}'

```

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/common_utils/src/types.rs [hunk 1]
```diff
@@ -738,6 +739,80 @@ impl Url {
...
+
+            if self.is_private_ip(&ip) {
+                return Err("Private URL not allowed.".to_string());
+            }
+        }
+
+        Ok(())
+    }
+
+    fn is_private_ip(&self, ip: &IpAddr) -> bool {
```
Comment (su-shivanshmathur) [inline crates/common_utils/src/types.rs]: Why is this needed ?

File: crates/common_utils/src/types.rs [hunk 2]
```diff
@@ -738,6 +739,80 @@ impl Url {
             .clone();
         Self(url)
     }
+
+    /// Validate the return URL provided in the ReportRequest payload
```
Comment (su-shivanshmathur) [inline crates/common_utils/src/types.rs]: What will we do with URL validation ?

**Review by su-shivanshmathur — CHANGES_REQUESTED** [review]

File: crates/common_utils/src/types.rs [hunk 3]
```diff
@@ -738,6 +739,80 @@ impl Url {
...
+        }
+
+        // Get host
+        let host = parsed
+            .host_str()
+            .ok_or_else(|| "URL must have a host.".to_string())?;
+
+        // Block localhost and common internal hostnames
+        let lowercase_host = host.to_lowercase();
+        let blocked_hosts = ["localhost", "127.0.0.1", "::1", "0.0.0.0"];
```
Comment (su-shivanshmathur) [inline crates/common_utils/src/types.rs]: Is this needed ?
Can this be config which can be set on server start?

File: crates/router/src/analytics.rs [hunk 4]
```diff
@@ -2189,6 +2204,11 @@ pub mod routes {
                             })
                             .ok_or(AnalyticsError::MissingEmail)?;
 
+                        if let Some(ref url) = payload.return_url {
```
Comment (su-shivanshmathur) [inline crates/router/src/analytics.rs]: Possible to shift these to a common validate function ?

**Review by su-shivanshmathur — CHANGES_REQUESTED** [review]

File: crates/common_utils/src/types.rs (hunk 2 above)
Comment (rohanm13) [reply crates/common_utils/src/types.rs]: Validating in the application layer will allow us to reject invalid or restricted IPs with 400 bad request. This prevents us from unnecessarily spinning up a Lambda worker, executing expensive database queries, and generating a report. It saves compute costs and gives the merchant immediate feedback.

File: crates/common_utils/src/types.rs (hunk 3 above)
Comment (rohanm13) [reply crates/common_utils/src/types.rs]: We need to validate - if the provided url parameter in the request is not an internal IP like localhost (127.0.0.1).
Similarly, we can validate against other internal IPs.

File: crates/common_utils/src/types.rs (hunk 1 above)
Comment (rohanm13) [reply crates/common_utils/src/types.rs]: We will block the internal or private IPs at the application layer itself, instead of invoking AWS lambda with an incorrect return_url. This prevents unnecessary spinning of lambda worker. It saves compute costs and gives the merchant immediate feedback to provide a valid return_url.

File: crates/common_utils/src/types.rs [hunk 5]
```diff
@@ -738,6 +739,33 @@ impl Url {
...
+            return Err("Invalid URL scheme. The scheme must be HTTPS.".to_string());
+        }
+
+        // Get host
+        let host = parsed
+            .host_str()
+            .ok_or_else(|| "URL must have a host.".to_string())?;
+
+        // DNS resolution and IP validation
+        let host_with_port = format!("{}:443", host);
```
Comment (su-shivanshmathur) [inline crates/common_utils/src/types.rs]: what if you get is another cloudfront url and not socket ?

File: crates/router/src/analytics_validator.rs [hunk 6]
```diff
@@ -22,3 +22,12 @@ pub async fn request_validator(
 
     Ok(ex_rates)
 }
+
+pub fn validate_report_request(request: &ReportRequest) -> Result<(), AnalyticsError> {
+    if let Some(ref url) = request.return_url {
+        url.verify_https_scheme()
+            .map_err(AnalyticsError::InvalidReturnUrl)?;
+    }
+
```
Comment (tsdk02) [inline crates/router/src/analytics_validator.rs]: Can use match statement here with Some(url) and None arms, better to avoid using `if let Some(x) = option(x)` patterns

File: crates/router/src/analytics.rs [hunk 7]
```diff
@@ -1894,6 +1894,7 @@ pub mod routes {
                         (user_email, payload.emails)
                     }
                     None => {
+                        validate_report_request(&payload)?;
```
Comment (tsdk02) [inline crates/router/src/analytics.rs:1795]: Are we adding this feature for the cases where these endpoints are hit programmatically, and not when user triggers it via dashboard?

File: crates/api_models/src/analytics.rs [hunk 8]
```diff
@@ -150,6 +154,8 @@ pub struct RefundDistributionBody {
 pub struct ReportRequest {
     pub time_range: TimeRange,
     pub emails: Option<Vec<Email>>,
+    #[serde(default)]
```
Comment (tsdk02) [inline crates/api_models/src/analytics.rs]: `default` flag can be removed since this is an optional field

File: crates/common_utils/src/types.rs [hunk 9]
```diff
@@ -763,6 +763,52 @@ where
...
+
+    /// Wrap a url::Url in MerchantWebhookUrl type
+    pub fn wrap(url: url::Url) -> Self {
+        Self(url)
+    }
+
+    /// Get the inner url::Url
+    pub fn into_inner(self) -> url::Url {
+        self.0
+    }
```
Comment (tsdk02) [inline crates/common_utils/src/types.rs]: Where are we using these?

**Comment by tsdk02** [comment]
Only checking for http and https scheme validation is sufficient? Are we not doing other validations?

File: crates/common_utils/src/types.rs [hunk 10]
```diff
@@ -763,6 +763,52 @@ where
     }
 }
 
+#[derive(
+    Debug, serde::Deserialize, serde::Serialize, Clone, PartialEq, Eq, Hash, ToSchema, PartialOrd,
+)]
+/// This domain type is specifically for merchant webhook URLs with validation
+pub struct MerchantWebhookUrl(url::Url);
+
+impl MerchantWebhookUrl {
```
Comment (su-shivanshmathur) [inline crates/common_utils/src/types.rs]: Move this to analytics please

**Review by su-shivanshmathur — CHANGES_REQUESTED** [review]

File: crates/common_utils/src/types.rs (hunk 9 above)
Comment (rohanm13) [reply crates/common_utils/src/types.rs]: These methods provide the standard newtype wrapper interface for `MerchantWebhookUrl`. While not all are currently used, they follow Rust conventions and provide a complete API for future use cases where we may need to extract or wrap the inner `url::Url` type.

File: crates/router/src/analytics.rs (hunk 7 above)
Comment (rohanm13) [reply crates/router/src/analytics.rs:1795]: yes, this feature is for the cases when endpoints are hit programmatically with API key based auth.

**Comment by rohanm13** [comment]
> Only checking for http and https scheme validation is sufficient? Are we not doing other validations?

@tsdk02 Discussed with @su-shivanshmathur. He told to remove those changes. Those are infra level checks.

**Review by su-shivanshmathur — DISMISSED** [review]

**Review by tsdk02 — DISMISSED** [review]

**Review by VenuMadhav2541 — DISMISSED** [review]

**Review by ShankarSinghC — APPROVED** [review]

**Review by ShankarSinghC — DISMISSED** [review]

**Review by VenuMadhav2541 — APPROVED** [review]

**Review by ShankarSinghC — APPROVED** [review]
